src/languages/internal/languagesservice.cpp

```diff
@@ -53,470 +59,340 @@ static QString downloadingStatusTitle()
-
-    return result;
+    return true;
 }
 
-RetCh<LanguageProgress> LanguagesService::update(const QString& languageCode)
+void LanguagesService::loadLanguages()
 {
-    RetCh<LanguageProgress> result;
-
-    if (m_operationsHash.contains(languageCode)) {
-        if (m_operationsHash[languageCode].type != OperationType::Install) {
-            result.ret = make_ret(Err::ErrorAnotherOperationStarted);
-            return result;
-        }
-
-        result.ret = make_ret(Err::NoError);
-        result.ch = *m_operationsHash[languageCode].progressChannel;
-        return result;
+    RetVal<ByteArray> languagesJson = fileSystem()->readFile(configuration()->builtinLanguagesJsonPath());
+    if (!languagesJson.ret) {
+        LOGE() << "Failed to read languages.json: " << languagesJson.ret.toString();
+        return;
     }
 
-    async::Channel<LanguageProgress>* languageProgressStatus = new async::Channel<LanguageProgress>();
-    result.ch = *languageProgressStatus;
-    result.ret = make_ret(Err::NoError);
-    m_operationsHash.insert(languageCode, Operation(OperationType::Install, languageProgressStatus));
-
-    async::Channel<Ret>* languageFinishChannel = new async::Channel<Ret>();
-    languageFinishChannel->onReceive(this, [this, languageCode, languageProgressStatus](const Ret& ret) {
-        if (!ret) {
-            LOGE() << ret.toString();
-            return;
-        }
-
-        closeOperation(languageCode, languageProgressStatus);
-    }, Asyncable::AsyncMode::AsyncSetRepeat);
-
-    QtConcurrent::run(this, &LanguagesService::th_update, languageCode, languageProgressStatus, languageFinishChannel);
-
-    return result;
-}
-
-Ret LanguagesService::uninstall(const QString& languageCode)
-{
-    LanguagesHash languagesHash = languages().val;
-
-    if (!languagesHash.contains(languageCode)) {
-        return make_ret(Err::ErrorLanguageNotFound);
+    QJsonParseError err;
+    QJsonDocument jsonDoc = QJsonDocument::fromJson(languagesJson.val.toQByteArrayNoCopy(), &err);
+    if (err.error != QJsonParseError::NoError || !jsonDoc.isObject()) {
+        LOGE() << "Failed to parse languages.json: " << err.errorString();
+        return;
     }
 
-    Ret remove = removeLanguage(languageCode);
-    if (!remove) {
-        return remove;
-    }
+    QJsonObject languagesObject = jsonDoc.object();
+    for (auto it = languagesObject.begin(); it != languagesObject.end(); ++it) {
+        Language lang;
+        lang.code = it.key();
+        lang.name = it.value().toString();
 
-    QString currentLanguageCode = currentLanguage().val.code;
-    if (languageCode == currentLanguageCode) {
-        resetLanguageToSystemLanguage();
+        m_languagesHash.insert(it.key(), lang);
     }
 
-    return make_ret(Err::NoError);
+    m_placeholderLanguage.code = PLACEHOLDER_LANGUAGE_CODE;
+    m_placeholderLanguage.name = "«Placeholder translations»";
 }
 
 void LanguagesService::setCurrentLanguage(const QString& languageCode)
 {
```

Comment:
There should be a check to make sure that the current language actually changed:

```
if (languageCode == currentLanguageCode()) {
    return;
}

...
```